src/task.js

```diff
@@ -118,7 +119,7 @@ function setupTask(fn, taskState, taskStateSchemaKeys, opts) {
      * Assigns the given properties to the task.
      * E.g. task.setState({ state: 'resolved', result: 1337 })
      */
-    setState: action(opts => {
+    setState: action(opts.wrappedFnName, opts => {
```

Comment:
Rather do it here than tack it on to `opts`, as that would require a typings update too.

```suggestion
    setState: action(fn.name || 'unnamed function', opts => {
```

Reply:
@jeffijoe that was my first thought as well; however, at least in my testing, fn has already been replaced with the task wrapper at the point and thus fn.name points to the task wrapper function instead of the original function that is being wrapped in the task.

I'm open to a cleaner solution as this wasn't my first preference because it is a little messy looking, but it seemed a slightly more streamlined way to get the data their without multiple new arguments being passed through the functions. And adding a const seemed inappropriate for something this minor.

Is there a way to reference the wrapped function from the wrapper function without delving into using prototypes?

Reply:
Actually, why are we even setting it on the task's `setState`? 😮

Reply:
To what are you referring? The first parameter of an action needs to either be the named function or a string name of the function passed as the second parameter in order to compatible with mobx-logger.

Reply:
Yes but setState is not the fn, which is where you are setting it

Reply:
Perhaps you want to use it like:

```js
@task
@action('name')
async fn() {}
```

Reply:
That works with decorators, does that work if you are using the decorate utility? *Sorry if this complicates things, trying to maintain full compatibility*

Doing something like this:
```
import React from 'react'
import { observable, action, decorate } from 'mobx'
import { task } from 'mobx-task'

class TestClass {
     testVar = null

     async runTestAction() {
     }

     async runTestTask() {
     }
}

decorate(TestClass, {
    testVar: observable,
    runTestAction: action,
    runTestTask: [task, action('runTestTaskName')] // or just runTestTask: [task, action]
})

export const testClass = new TestClass()
```

returns this error: `[mobx] @action cannot be used with getters`

It doesn't appear that task and action can be combined in the decorator utility...?

Reply:
Try reversing the order; order matters.

Reply:
Just tried this:

```
import React from 'react'
import { observable, action, decorate } from 'mobx'
import { task } from 'mobx-task'

class TestClass {
     testVar = null

     async runTestAction() {
     }

     async runTestTask() {
     }
}

decorate(TestClass, {
    testVar: observable,
    runTestAction: action,
    runTestTask: [action('runTestTaskName'), task]
})

export const testClass = new TestClass()
```
It runs without errors, but it does not name the action/task:

Logged result is: `12:32:08.746    ACTION  task.<unnamed action>()`

Reply:
The only `action` used in `mobx-task` is for `setState`; is that the one being logged? Is it logging twice? (one for pending, one for resolved)

Reply:
It is only being logged once - regardless of whether the function is decorated with `task` or `task.resolved` only one log entry is created per task/action called.

Reply:
But `mobx-task` does not use `action` for the `fn` you pass in? 😮

Reply:
I apologize - I was definitely not as clear as I should have been - mobx-logger is just handling the formatted output to console essentially. The `<unnamed action>` is actually directly from mobx core.

`mobx/src/api/action.ts`

Starting around line 56:
```
export const action: IActionFactory = function action(arg1, arg2?, arg3?, arg4?): any {
    // action(fn() {})
    if (arguments.length === 1 && typeof arg1 === "function")
        return createAction(arg1.name || "<unnamed action>", arg1)
    // action("name", fn() {})
    if (arguments.length === 2 && typeof arg2 === "function") return createAction(arg1, arg2)

    // @action("name") fn() {}
    if (arguments.length === 1 && typeof arg1 === "string") return namedActionDecorator(arg1)

    // @action fn() {}
    if (arg4 === true) {
        // apply to instance immediately
        addHiddenProp(arg1, arg2, createAction(arg1.name || arg2, arg3.value, this))
    } else {
        return namedActionDecorator(arg2).apply(null, arguments as any)
    }
} as any

action.bound = boundActionDecorator as any
```

As far as I can tell when mobx-task calls the `setState` in `setupTask` it calls upon `action` from mobx core which initializes the `IActionFactory`. This has fn returned into the `action` thus provisioning the `IActionFactory`.
```
   setState: action(opts => {
      Object.assign(taskState, opts)
      return fn
    }),
```

I tried getting `fn.name` directly inside of `setupTask` but the function has already been modified at that point and thus `fn.name` is simply "task" instead of the name of the originating function.

Reply:
Maybe I am misunderstanding something, but it sounds to me like you are trying to name the `fn`, while what the PR is doing is naming the `setState`, which does not call `fn` itself.

Does that make sense?

Reply:
ahh - I understand your question now. So it is definitely odd. I'm trying to apply the name to the action IActionFactory in MOBX ultimately since that is what is what is used for logging/identification/audit of MOBX actions - and thusly tasks.

As a result that has to be done when `action` is called - which in `task.js` is only referenced in setState.

For example if in the set state you do something like - `setState: action('MyActionName', opts => {` 'MyActionName' will get applied to the action/IActionFactory in MOBX thus logging as `task.MyActionName()` having been called with call details via mobx-logger.

Is it possible that the `action` definition itself in `task.js` is better suited to live outside of the `setState`?

It definitely seems like there should be a simple way to accomplish this without going "off book".

Reply:
The reason `setState` is an action is because it modifies *internal observable state* (pending, result, error, etc), so if you want your actual task to be an action, you need to wrap the input to `task` in an action. But since it's async, I don't see a point in doing so as actions are inherently sync.

Reply:
I agree, the `action` should be in `setState`, so we circle back around to the best way to pass a first parameter into the `setState` `action` before the the function that is passed as the second parameter.

What if instead of altering options, we do something like this:

**Change line 82 to be:**
```javascript
setupTask(task, taskState, taskStateSchemaKeys, opts, fn.name || 'undefinedTask')
```

**Change line 89 to be:**
```javascript
function setupTask(fn, taskState, taskStateSchemaKeys, opts, taskName)
```

**Change line 121 to be:**
```javascript
setState: action(taskName, opts => {
```

It is an extra line modified, but it prevents altering opts, just adds an extra parameter to task setup - which is probably where it should be anyway since your are essentially defining an attribute of the task.

Reply:
I still don't understand the benefit to making `setState` a named action other than maybe `action('setState', () => ...)`. Have you tried making the modifications to a local installation to see if you get the results you want?

Reply:
Yes, I'm testing all this against a local react app.

This is what it looks like with current released version of `mobx-task`:
![Image of CurrentMobxTaskRelease](https://imgur.com/IKNs2Lp)

This is what it looks like with any version of the "name" patch - it includes an expanded *standard* MOBX `action` to see the current difference.
![Image of MobxTaskWithNamePatch](https://imgur.com/iq8eabz)

Reply:
Could you expand the args on the last one? Because I think `setState` is what is being logged, which happens inside the task but isn't the actual `fn` itself.

Reply:
Expanded output:
[https://imgur.com/LHocCoK](https://imgur.com/LHocCoK)

Reply:
Yeah so my assumption seems correct; it's logging the `setState` call, not the actual `fn` itself. So the correct way would be `setState: action('setState', () => ...)`

Reply:
This is currently how `mobx-logger` is outputting the action event:

```javascript
const logAction = (ev) => {
    if (!isLoggingEnabled(ev)) {
        return;
    }

    console.groupCollapsed('%c%s  %s  %s.%s()', style('#000'), now(), padStart('ACTION', 8), ev.object.name || ev.object, ev.name);
    console.log('%cFunction %o', style('#777'), ev.fn);
    console.log('%cArguments %o', style('#777'), ev.arguments);
    console.log('%cTarget %o', style('#777'), ev.object);
    console.log('%cEvent %o', style('#777'), ev);
    console.groupEnd();
};
```
This is how `mobx-logger` is getting initialized:
```javascript
export const enableLogging = (options = defaultOptions) => {
    const predicate = options.predicate || defaultOptions.predicate;
    if (predicate() === true) {
        return spy(ev => log(ev, options));
    }
    return () => void(0);
};
```
Since we want to log the actual `task` occurring - not just a `setState` then, does something else entirely need to be added? I guess the root question is why is a `event` only firing on setState and not when the actual `task` is being executed...

Reply:
If you wrap it like `myFunction = task(action(myFunction))` or `@task @action myFunction() {}` then it should do what you want?

Reply:
The thing is though, a Task isn't really an action; it's async so it shouldn't be. You want to use actions *inside* the Task though.

Reply:
The `action` part is irrelevant I think - it does not have to be an `action` or `sync`.

I guess the reality of the situation is, if you are debugging an app/mobx, wouldn't we want a log message when the *async* `task` is fired and when it completes/resolves. This seems like is what would really be helpful.

`mobx-logger` currently has multiple "log functions" `action`, `reaction`, etc. however they all require an event to be fired. [https://github.com/winterbe/mobx-logger/blob/master/src/log.js](https://github.com/winterbe/mobx-logger/blob/master/src/log.js)

So I guess the question is, why isn't an event firing when the `task` is fired/resolved. Why is it only firing on `setState`. Do async `tasks` not fire events? Seems like something should know about the event happening.

Reply:
Because `fn` in `task(fn)` is not wrapped in an action, and so therefore MobX can't spy on it. If you wrap it in an action, then it should log it whenever it runs.

Reply:
That makes sense - so why then is `setState` utilizing a *mobx* `action`?

Reply:
Because a Task has internal state that is observable, and to update observables you need to do it in an action.

Reply:
Ahh ok - because they are being mixed. Got it. I will adjust this PR to just be the `"setState"` added to the `action(`. I will work separately on figuring out a better solution for general task capture into log.

Reply:
Changes made to PR.

Reply:
Released as 1.0.4, thanks!

Reply:
Thanks! I was hoping to implement a custom `I*RunInfo` with `spyReportStart` and `spyReportEnd` the same way that `action` implements its `event` record creation. Sadly, this was apparently removed from the API between MOBX 3 and 4. So guess for now this is the best that can be without getting messy.